bittide/src/Bittide/Transceiver/WordAlign.hs

```diff
@@ -0,0 +1,178 @@
+--
+-- ..might be received on the other end as:
+--
+-- > -----------------------------------------------
+-- > | .. .. B3 B2 | B1 B0 B3 B2 | B1 B0 B3 B2 | ...
+-- > -----------------------------------------------
+
+-- > -----------------------------------------------
+-- > | .. .. B0 B1 | B2 B0 B3 B2 | B1 B0 B3 B2 | ...
+-- > -----------------------------------------------
+--
+-- or by any other shift (or none at all!). This module provides utilities to end
+-- up with a word aligned stream. The basic idea is that, while \"booting\" the
+-- connection the MSB of each byte is reserved, where an asserted MSB indicates
+-- the start of a word. The other bits can be used to detect a valid connection
+-- through PRBS streams.
+--
+-- TODO: Remove this module in favor of retry logic. That is, we can keep
+--       resetting the transceivers until they're aligned. Its unclear how exactly
+--       this should work though.
+module Bittide.Transceiver.WordAlign
+  ( alignBytesFromMsbs
+  , align
+
+  -- * Convenience functions
+  , alignSymbol
+  , splitMsbs
+  , joinMsbs
+  , dealignBytes
+  ) where
+
+import Clash.Prelude
+
+import Bittide.SharedTypes (Bytes)
+import Clash.Explicit.Prelude (noReset)
+import Data.Bifunctor (Bifunctor (bimap))
+import Data.Maybe (fromMaybe)
+import Data.Tuple.Extra (curry3)
+
+data State n m = State { prev :: BitVector (n * m), offset :: Index n }
+  deriving (Show, Generic, ShowX, NFDataX)
+
+defState :: (KnownNat n, KnownNat m) => State n m
+defState = State{prev=0, offset=0}
+
+alignSymbol :: forall n . (KnownNat n, 1 <= n) => BitVector n
+alignSymbol = bit (natToNum @(n - 1))
+
+-- | Split the MSBs of a 'BitVector's \"bytes\" into a 'BitVector' of MSBs and a
+-- 'BitVector' of the remaining bits.
+splitMsbs ::
+  forall nBytes byteWidth .
+  ( KnownNat nBytes
+  , KnownNat byteWidth
+  , 1 <= byteWidth
+  ) =>
+  BitVector (nBytes * byteWidth) ->
+  ( BitVector nBytes
+  , BitVector (nBytes * (byteWidth - 1))
+  )
+splitMsbs =
+    bimap pack pack
+  . unzip
+  . unpack @(Vec nBytes (Bit, BitVector (byteWidth - 1)))
+
+-- | Opposite of 'splitMsbs'.
+joinMsbs ::
+  forall nBytes byteWidth .
+  ( KnownNat nBytes
+  , KnownNat byteWidth
+  , 1 <= byteWidth
+  ) =>
+  BitVector nBytes ->
+  BitVector (nBytes * (byteWidth - 1)) ->
+  BitVector (nBytes * byteWidth)
+joinMsbs msbs bvs = pack $
+  zip
+    (unpack @(Vec nBytes Bool) msbs)
+    (unpack @(Vec nBytes (BitVector(byteWidth - 1))) bvs)
+
+-- | Specialized version of 'align' that works on 'BitVector' and assumes that
+-- the alignment bits are stored in the MSBs of the bytes.
+alignBytesFromMsbs ::
+  forall n dom .
+  ( HiddenClock dom
+  , KnownNat n
+  , 1 <= n
+  ) =>
+  -- | Freeze alignment
+  Signal dom Bool ->
+  -- | Data with alignment bits in the byte's MSBs
+  Signal dom (Bytes n) ->
+  -- | 'Nothing' until an alignment symbol has been observed. 'Just' if an alignment
+  -- symbol has been observed.
+  Signal dom (Bytes n)
```

Comment:
According to the documentation it seems that some `Maybe a` gets returned, but that's actually not the case. Is the documentation maybe not up to date?

Reply:
Yeah.. that documentation was a mess. What do you think of the current state?

Reply:
Makes more sense now.